### PR TITLE
[MBL-17617][Parent] Parent announcement details screen

### PR DESCRIPTION
**Pull request by andrasmaczak**
Test plan: Test screen with both course announcements and global announcements

refs: MBL-17617
affects: Parent
release note: none

## Screenshots

Screenshot_20241014_154002

## Checklist

- [x] Run E2E test suite
- [x] Tested in dark mode
- [x] Tested in light mode

### CODE REVIEW

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+    uiState: AnnouncementDetailsUiState,
+    actionHandler: (AnnouncementDetailsAction) -> Unit,
+    navigationActionClick: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    Scaffold(
+        backgroundColor = colorResource(id = R.color.backgroundLightest),
+        topBar = {
+            CanvasThemedAppBar(
+                title = uiState.pageTitle
```
Comment (tamaskozmer) [inline apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt:94]: I see that the `pageTitle` logic is already implemented in the ViewModel so we don't really need to handle cases here. You can just make it an empty string by default in the UiState so you won't need to handle null cases. We usually add default values in the UiState for basic types so we don't clutter the UI layer with additional logic.

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+                                color = colorResource(id = R.color.textDarkest),
+                                fontSize = 24.sp
+                            )
+                        )
+                        Text(
+                            text = DateHelper.getDateAtTimeString(
+                                LocalContext.current,
+                                R.string.alertDateTime,
+                                uiState.postedDate
+                            ) ?: "",
```
Comment (tamaskozmer) [inline apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt:130]: In cases like these where it wouldn't make sense to add an empty Date() we could just keep the nullable date in the UiState. You can use `orEmpty()` though for the string.

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+                }
+            )
+            Box(modifier = modifier.pullRefresh(pullRefreshState)) {
+                Column(
+                    modifier = Modifier
+                        .fillMaxSize()
+                        .verticalScroll(rememberScrollState())
+                ) {
+                    Column(
+                        modifier = modifier.padding(16.dp)
```
Comment (tamaskozmer) [inline apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt:116]: I think this padding causes the issue with the WebView. I would move this one level deeper.

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt [hunk 4]
```diff
@@ -0,0 +1,184 @@
...
+import com.instructure.pandautils.compose.composables.CanvasThemedAppBar
+import com.instructure.pandautils.compose.composables.ComposeCanvasWebViewWrapper
+import com.instructure.pandautils.compose.composables.ErrorContent
+import com.instructure.pandautils.compose.composables.Loading
+import com.instructure.pandautils.features.inbox.utils.AttachmentCard
+import com.instructure.pandautils.features.inbox.utils.AttachmentCardItem
+import com.instructure.pandautils.features.inbox.utils.AttachmentStatus
+
+@Composable
+fun AnnounceDetailsScreen(
```
Comment (tamaskozmer) [inline apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt:45]: There should be a preview for every screen.

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/list/AlertsViewModel.kt [hunk 5]
```diff
@@ -126,7 +127,11 @@ class AlertsViewModel @Inject constructor(
         when (action) {
             is AlertsAction.Navigate -> {
                 viewModelScope.launch {
-                    _events.send(AlertsViewModelAction.Navigate(action.route))
+                    _events.send(AlertsViewModelAction.Navigate(
```
Comment (tamaskozmer) [inline apps/parent/src/main/java/com/instructure/parentapp/features/alerts/list/AlertsViewModel.kt:130]: I would refactor this logic and move the conditions from the Fragment into the ViewModel and would create separate actions for navigating to different alerts.

**Review by tamaskozmer — CHANGES_REQUESTED** [review]
QA findings:

- [ ] I think the Toolbar title should be "Institution Announcement" like in the current app. We should also change this text on the alert list as well. Or was this intentional?
- [ ] I would remove the padding of the WebView since it already has an inside padding and it doesn't look aligned this way.
- [ ] Pull to refresh shouldn't clear the screen and show the initial screen loading. Also we shouldn't clear the already present data on pull to refresh, instead we should just show a Snackbar or Toast as an error message.
- [ ] I think the error screen and loading screen should also include the Toolbar. It looks weird without it.
- [ ] The status bar does not have the correct color after a config change.
- [ ] The back navigation is broken after a config change, it navigates back to the courses list screen.

Not in the scope of this ticket:
- [ ] When refreshing alerts with pull to refresh and we have new items they are not shown because the screen won't scroll up automatically.

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by andrasmaczak** [comment]
> * I think the Toolbar title should be "Institution Announcement" like in the current app. We should also change this text on the alert list as well. Or was this intentional?

This was intentional.

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Review by tamaskozmer — APPROVED** [review]

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsViewModel.kt [hunk 6]
```diff
@@ -37,26 +37,39 @@ class AnnouncementDetailsViewModel @Inject constructor(
     val uiState = _uiState.asStateFlow()
 
     init {
+        _uiState.update {
+            it.copy(isLoading = true)
+        }
         loadData()
     }
 
+    var first = true
```
[the remark this answers is not in the document]
Comment (andrasmaczak) [reply apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsViewModel.kt:46]: Added for testing, will remove

File: apps/parent/src/test/java/com/instructure/parentapp/features/alerts/list/AlertsViewModelTest.kt [hunk 7]
```diff
@@ -642,10 +642,80 @@ class AlertsViewModelTest {
         }
 
         assertEquals(
-            AlertsViewModelAction.Navigate(
+            AlertsViewModelAction.NavigateToCourseAnnouncement(
```
[the remark this answers is not in the document]
Comment (andrasmaczak) [reply apps/parent/src/test/java/com/instructure/parentapp/features/alerts/list/AlertsViewModelTest.kt:645]: Done

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Review by kdeakinstructure — APPROVED** [review]
QA 👍

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+fun AnnouncementDetailsScreen(
+    uiState: AnnouncementDetailsUiState,
+    actionHandler: (AnnouncementDetailsAction) -> Unit,
+    navigationActionClick: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    CanvasTheme {
+        val snackbarHostState = remember { SnackbarHostState() }
+        val errorMessage = stringResource(id = R.string.errorLoadingAnnouncement)
+        LaunchedEffect(uiState.showErrorSnack) {
```
Comment (domonkosadam) [inline apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt:63]: Wouldn't be better to handle these Snackbars with Channel and Flows events like on the other new screens?

**Review by domonkosadam — APPROVED** [review]
QA + 1

File: apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt (hunk 8 above)
Comment (tamaskozmer) [reply apps/parent/src/main/java/com/instructure/parentapp/features/alerts/details/AnnouncementDetailsScreen.kt:63]: I think this is the correct way for Compose. What you mention is used on legacy screens.

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Review by tamaskozmer — APPROVED** [review]

**Comment by inst-danger** [comment]
<p>Student Install Page</p>

**Comment by inst-danger** [comment]
<p>Teacher Install Page</p>

**Comment by inst-danger** [comment]
<p>Native Parent Install Page</p>

**Pull request merged**